### PR TITLE
@ #33 | create provider manager

### PR DESCRIPTION
**Pull request by hieptranquoc**
connect #33 
please check this PR. Thank you!

### CODE REVIEW

**Comment by hieptranquoc** [comment]
@hoatle @datphan, please help me review this PR!

File: lib/teracy-dev-core.rb [hunk 1]
```diff
@@ -26,10 +25,8 @@ def self.init
     TeracyDev.register_configurator(TeracyDevCore::Config::SSH.new)
     TeracyDev.register_configurator(TeracyDevCore::Config::SyncedFolders.new)
     TeracyDev.register_configurator(TeracyDevCore::Config::Vgrant.new)
-    TeracyDev.register_configurator(TeracyDevCore::Config::VirtualBoxProvider.new)
-    TeracyDev.register_configurator(TeracyDevCore::Config::VMwareProvider.new) # Use for vmware provider.
+    TeracyDev.register_configurator(TeracyDevCore::Config::Provider.new)
     TeracyDev.register_configurator(TeracyDevCore::Config::WinRM.new)
     TeracyDev.register_configurator(TeracyDevCore::Config::WinSSH.new)
   end
```
Comment (hoatle) [inline lib/teracy-dev-core.rb:45]: Let's add public api for the TeracyDevCore:

TeracyDevCore.register_provider(provider_type: String, provider: Provider)

TeracyDevCore.register_provider is the proxy to the TeracyDevCore::Providers::Manager

Providers::Manager is a singleton class (the same with kernel configManager or processorManager) that holds the registered providers (registry = {"x_type": [x_provider_impl]})
Providers::Manager allows other classes to access the registry

File: lib/teracy-dev-core/config/provider.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+require 'teracy-dev/config/configurator'
+
+require_relative '../providers/manager'
+
+module TeracyDevCore
+  module Config
+    class Provider < TeracyDev::Config::Configurator
+
+      attr_reader :providerManager
```
Comment (hoatle) [inline lib/teracy-dev-core/config/provider.rb]: remove this

File: lib/teracy-dev-core/config/provider.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+      def configure_node(settings, config)
+        providers_settings = settings['providers'] ||= []
+        @logger.debug("providers_settings: #{providers_settings}")
+
+        @providerManager = TeracyDevCore::Providers::Manager.new
```
Comment (hoatle) [inline lib/teracy-dev-core/config/provider.rb]:
```ruby
registry = Providers::Manager.registry

providers_settings.each do |provider_settings|
  next if !provider_settings['enabled']
  providers = registry(provider_settings['type'])
  # TODO: check if there is any provider implementation registered
  # if not, add warning and next
  providers.each do |provider|
    provider.configure(provider_settings, config)
  end
end
```

File: lib/teracy-dev-core/providers/manager.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      def initialize
+        if !!@@instance
+          raise "TeracyDevCore::Providers::Manager can only be initialized once"
+        end
+
+        @@instance = self
+        @logger = TeracyDev::Logging.logger_for(self.class.name)
+      end
+
+      def register(provider_type)
```
Comment (hoatle) [inline lib/teracy-dev-core/providers/manager.rb]:
```ruby
      def register(provider_type, provider_impl)
        # make sure provider_impl must implement the configure method (Provider interface)
        registry[provider_type] = [provider_impl] if registry[provider_type] is nil
        registry[provider_type] << provider_impl
      end
```

File: lib/teracy-dev-core/providers/manager.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        case provider_type
+        when 'virtualbox'
```
Comment (hoatle) [inline lib/teracy-dev-core/providers/manager.rb]: we never know beforehand the type of providers, this is considered hard-code and not understanding the flow.

File: lib/teracy-dev-core/providers/provider.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+module TeracyDevCore
+  module Providers
+    # the base class for Provider to extend
+    class Provider
+      def initialize
+        @logger = TeracyDev::Logging.logger_for(self.class.name)
+      end
+
+      # provider must implement this to execute actual configuration
+      def configure_provider(settings, config)
```
Comment (hoatle) [inline lib/teracy-dev-core/providers/provider.rb]: def configure(provider_settings, node_config)

File: lib/teracy-dev-core/providers/virtualbox.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
+require_relative 'provider'
+
+module TeracyDevCore
+  module Providers
+    # see: https://www.vagrantup.com/docs/virtualbox/
+    class VirtualBox < TeracyDevCore::Providers::Provider
+
+      def configure_provider(settings, config)
```
Comment (hoatle) [inline lib/teracy-dev-core/providers/virtualbox.rb]: configure(provider_settings, node_config)

File: lib/teracy-dev-core/providers/vmware.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
+require_relative 'provider'
+
+module TeracyDevCore
+  module Providers
+    # see: https://www.vagrantup.com/docs/vmware/
+    class VMware < TeracyDevCore::Providers::Provider
+
+      def configure_provider(settings, config)
```
Comment (hoatle) [inline lib/teracy-dev-core/providers/vmware.rb]: def configure(provider_settings, node_config)

File: lib/teracy-dev-core.rb [hunk 9]
```diff
@@ -26,10 +25,8 @@ def self.init
     TeracyDev.register_configurator(TeracyDevCore::Config::SSH.new)
     TeracyDev.register_configurator(TeracyDevCore::Config::SyncedFolders.new)
     TeracyDev.register_configurator(TeracyDevCore::Config::Vgrant.new)
-    TeracyDev.register_configurator(TeracyDevCore::Config::VirtualBoxProvider.new)
-    TeracyDev.register_configurator(TeracyDevCore::Config::VMwareProvider.new) # Use for vmware provider.
+    TeracyDev.register_configurator(TeracyDevCore::Config::Provider.new)
     TeracyDev.register_configurator(TeracyDevCore::Config::WinRM.new)
     TeracyDev.register_configurator(TeracyDevCore::Config::WinSSH.new)
```
Comment (hoatle) [inline lib/teracy-dev-core.rb:44]:
```
self.register_provider("virtualbox", Providers::VirtualBoxProvider.new)
self.register_provider("vmware_desktop", Providers::VMwareProvider.new)
```

**Review by hoatle — CHANGES_REQUESTED** [review]
@hieptranquoc make sure to study and understand how everything is connected and works to implement the requirements precisely and  properly

File: lib/teracy-dev-core.rb (hunk 1 above)
Comment (hoatle) [reply lib/teracy-dev-core.rb:45]: @hieptranquoc check my updated comment here

**Comment by hieptranquoc** [comment]
@hoatle I've updated! please check again.

File: lib/teracy-dev-core/config/provider.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        registry = Providers::Manager.registry
+        providers_settings = settings['providers'] ||= []
+        @logger.debug("providers_settings: #{providers_settings}")
+
+        providers_settings.each do |provider_settings|
+          next if !provider_settings['enabled']
+          providers = registry[provider_settings['type']]
+
+          if providers.nil?
+            @logger.warn("#{provider_settings['type']} have not registered")
```
Comment (hoatle) [inline lib/teracy-dev-core/config/provider.rb]: `@logger.warn("not any provider with the type of #{provider_settings['type']} is registered")`

File: lib/teracy-dev-core/config/provider.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+      def configure_node(settings, config)
+        registry = Providers::Manager.registry
+        providers_settings = settings['providers'] ||= []
+        @logger.debug("providers_settings: #{providers_settings}")
+
+        providers_settings.each do |provider_settings|
+          next if !provider_settings['enabled']
+          providers = registry[provider_settings['type']]
+
+          if providers.nil?
```
Comment (hoatle) [inline lib/teracy-dev-core/config/provider.rb]: nil? or is an empty array?

File: lib/teracy-dev-core/providers/manager.rb [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+      @registry = {}
+
+      class << self
+        attr_reader :registry
+      end
+
+      def self.register(provider_type, provider_impl)
+        @registry[provider_type] = [provider_impl] if @registry[provider_type].nil?
+        @registry[provider_type] << provider_impl
+        @@logger.debug("provider: #{provider_type} registered")
```
Comment (hoatle) [inline lib/teracy-dev-core/providers/manager.rb]:
```
@@logger.debug("#{provider_impl} with type: #{provider_type} registered")
```

File: lib/teracy-dev-core/providers/manager.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+      @@logger = TeracyDev::Logging.logger_for(self)
+      @registry = {}
+
+      class << self
+        attr_reader :registry
+      end
+
+      def self.register(provider_type, provider_impl)
+        @registry[provider_type] = [provider_impl] if @registry[provider_type].nil?
```
Comment (hoatle) [inline lib/teracy-dev-core/providers/manager.rb]: validate provider_impl that must implement the configure method signature

**Review by hoatle — CHANGES_REQUESTED** [review]
@hieptranquoc pls check my comments

**Comment by hieptranquoc** [comment]
@hoatle I've updated! Please check again! Thank you!

File: lib/teracy-dev-core/config/provider.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+          providers = registry[provider_settings['type']]
+
+          if providers.nil? || providers.empty?
+            @logger.warn("not any provider with the type of #{provider_settings['type']} is registered")
+            next
+          end
+
+          providers.each do |provider|
+            provider.configure(provider_settings, config)
+            break
```
Comment (hoatle) [inline lib/teracy-dev-core/config/provider.rb]: why `break` here?

File: lib/teracy-dev-core/providers/manager.rb [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+      end
+
+      def self.register(provider_type, provider_impl)
+        if !provider_impl.respond_to?(:configure)
+          @logger.warn("provider #{provider_impl} must implement configure method, ignored")
+          return
+        end
+
+        @registry[provider_type] = [provider_impl] if @registry[provider_type].nil?
+        @registry[provider_type] << provider_impl
```
Comment (hoatle) [inline lib/teracy-dev-core/providers/manager.rb]: bug here:

```
[TeracyDevCore::Providers::Manager][DEBUG]: registry: {"virtualbox"=>[#<TeracyDevCore::Providers::VirtualBox:0x00000001020eaf80 @logger=#<Logger:0x00000001020eaf30 @level=0, @progname="TeracyDevCore::Providers::VirtualBox", @default_formatter=#<Logger::Formatter:0x00000001020eadf0 @datetime_format=nil>, @formatter=#<Proc:0x00000001020eab98@/Users/hoatle/teracy-dev/lib/teracy-dev/logging.rb:44>, @logdev=#<Logger::LogDevice:0x00000001020ead28 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x00000001020eacd8>>>>, #<TeracyDevCore::Providers::VirtualBox:0x00000001020eaf80 @logger=#<Logger:0x00000001020eaf30 @level=0, @progname="TeracyDevCore::Providers::VirtualBox", @default_formatter=#<Logger::Formatter:0x00000001020eadf0 @datetime_format=nil>, @formatter=#<Proc:0x00000001020eab98@/Users/hoatle/teracy-dev/lib/teracy-dev/logging.rb:44>, @logdev=#<Logger::LogDevice:0x00000001020ead28 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x00000001020eacd8>>>>], "vmware_desktop"=>[#<TeracyDevCore::Providers::VMware:0x0000000101266798 @logger=#<Logger:0x0000000101267440 @level=0, @progname="TeracyDevCore::Providers::VMware", @default_formatter=#<Logger::Formatter:0x00000001008578b0 @datetime_format=nil>, @formatter=#<Proc:0x00000001008541b0@/Users/hoatle/teracy-dev/lib/teracy-dev/logging.rb:44>, @logdev=#<Logger::LogDevice:0x00000001008571a8 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x0000000100856f50>>>>, #<TeracyDevCore::Providers::VMware:0x0000000101266798 @logger=#<Logger:0x0000000101267440 @level=0, @progname="TeracyDevCore::Providers::VMware", @default_formatter=#<Logger::Formatter:0x00000001008578b0 @datetime_format=nil>, @formatter=#<Proc:0x00000001008541b0@/Users/hoatle/teracy-dev/lib/teracy-dev/logging.rb:44>, @logdev=#<Logger::LogDevice:0x00000001008571a8 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x0000000100856f50>>>>]}
/Users/hoatle/teracy-dev/workspace/teracy-dev-core/lib/teracy-dev-core/providers/manager.rb:23:in `register'
```

**Comment by hieptranquoc** [comment]
@hoatle I've updated!

File: lib/teracy-dev-core/providers/manager.rb (hunk 15 above)
Comment (hoatle) [reply lib/teracy-dev-core/providers/manager.rb]: one provider_impl was registered into the registry twice, this is a bug

Comment (hoatle) [reply lib/teracy-dev-core/providers/manager.rb]: logging with debug should help, try to use the debug log where we should use to check the changes.

Comment (hoatle) [inline lib/teracy-dev-core/providers/manager.rb]:
```ruby
      def self.register(provider_type, provider_impl)
        if !provider_impl.respond_to?(:configure)
          @logger.warn("provider #{provider_impl} must implement configure method, ignored")
          return
        end

        if @registry[provider_type].nil?
          @registry[provider_type] = [provider_impl]
        else
          @registry[provider_type] << provider_impl
        end
        @@logger.debug("#{provider_impl} with type: #{provider_type} registered")
        @@logger.debug("registry: #{@registry}")
      end
```

**Comment by hieptranquoc** [comment]
@hoatle I've fixed, please check again! Thank you so so much!

**Review by hoatle — APPROVED** [review]
LGTM, merging, will do some minor clean up

**Pull request merged**